### PR TITLE
feat(ui): Add drawer with more details for each workflow in Workflow List

### PR DESCRIPTION
**Pull request by rbreeze**
Checklist:

* [ ] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this is a chore.
* [x] The title of the PR is (a) [conventional](https://www.conventionalcommits.org/en/v1.0.0/), (b) states what changed, and (c) suffixes the related issues number. E.g. `"fix(controller): Updates such and such. Fixes #1234"`.  
* [x] I've signed the CLA.
* [ ] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [ ] My builds are green. Try syncing with master if they are not. 
* [x] My organization is added to [USERS.md](https://github.com/argoproj/argo/blob/master/USERS.md).

Screen Shot 2020-06-02 at 5 10 16 PM

### CODE REVIEW

**Comment by alexec** [comment]
be very selective in the amount of data you query yes? there maybe >2k workflow listed

**Comment by rbreeze** [comment]
@alexec Is the data I've added so far too much? I discussed this a bit with @simster7  and those were the fields he suggested but I can remove some if needed. I also believe the drawer doesn't get rendered until a user clicks the SHOW button

**Comment by simster7** [comment]
> be very selective in the amount of data you query yes? there maybe >2k workflow listed

I believe the data has already been transmitted to the UI at this point, it's simply being surfaced now. Also, it might be a bit hard to see but the drawer is only shown – and therefore only rendered by React – if a user decides to display it.

**Comment by rbreeze** [comment]
> I believe the data has already been transmitted to the UI at this point, it's simply being surfaced now.

Yeah, each workflow object is already there in its entirety, I'm just displaying more of its properties in the drawer.

**Comment by simster7** [comment]
> We should not be transmitting the data of EVERY workflow that is loaded into this page. Only top line items. Can I confirm that if you open the drawer, then you are only loading the workflow at that point?

Sorry, not sure I understand what your concern is here. Could you help clarify?

If you're concerned about sending more information over-the-wire, then this PR does not send any new information that was not previously sent before. As of right now, we send Workflow objects in their entirety to the endpoint used by the timeline page. This PR simply exposes more of the information that is already being sent. If you're proposing we strip Workflow objects sent through this endpoint down to only the necessary fields to display them, then that would be out of scope of this PR.

If you're concerned about the UI having performance issues when loading or displaying the information that it has received, then React renders the information intelligently and will only do so if the user clicks the "Show" button.

**Comment by alexec** [comment]
@simster7 I'm confused - the screen shot in the description is of the _workflow list_ page - but you're talking about the _timeline page_ - I don't know what page that is - the only thing it might be is the _timeline tab_  fo the _workflow page_?

Can you confirm which page makes these changes?

**Comment by alexec** [comment]
Looking at the changes, this is the _workflow list page_. This page should not be listing all workflow details - only those relevant to the list. This is to reduce network I/O for long lists (1000+ workflows).

This should be achieved by line 26 in `workflow-service.ts`:

```
fields=metadata,items.metadata.name,items.metadata.namespace,items.status.phase,items.status.finishedAt,items.status.startedAt
```

If that is actually returning more information, then we have a bug that this PR has revealed. However, what I expect is happening is that events that come from `/workflow-events` contain complete workflows. If that is the case, this solution will probably not work for large numbers of workflows -  as you won't get that data and so they'll - missing the data you're rendering - I don't know how the changes handle that - but it could crash the UI.

If I'm correct, then can you make sure you're only getting the specific minimal data you need for this? Even adding labels to `WorkflowService.list` might be a lot.

Or, maybe remove the drawer if that data is missing.

**Comment by rbreeze** [comment]
That screenshot was a poor example because that workflow didn't contain a `message`. Here's an example of one with all the data that's being displayed in the drawer: 

Screen Shot 2020-06-03 at 10 56 18 AM

**Comment by simster7** [comment]
I see, thank you for clarifying.

> I'm confused - the screen shot in the description is of the workflow list page - but you're talking about the timeline page - I don't know what page that is - the only thing it might be is the timeline tab fo the workflow page?

You're correct, we're talking about the same page. I was using the term "timeline" that referred to the help string when hovering over the icon, sorry for the confusion
image

> This should be achieved by line 26 in workflow-service.ts

Thanks for that, I wasn't aware of that line. In that case it does seem like there is a bug and extra information is being provided as @rbreeze just mentioned. We should fix this in a separate PR.

For scope within this PR, I think it would be helpful to provide _some_ details in this Workflows list page via the drawer in question. We could either:

1. Choose some fileds that we think are important to include and add them to line 26 in workflow-service.ts to be included. In my opinion, "Message", "Conditions", and "Labels" should be included.

2. We can "lazy-load" the rest of the workflow information by performing an API call when opening the drawer.

I am indifferent to either solution as I think both work well. Thoughts? @alexec @rbreeze

**Comment by rbreeze** [comment]
@simster7 Both suggestions sound good, but lazy loading seems like the better solution to me because it would allow for us to display fields without compromise and without having to worry about server load or UI lag. Additionally, the delay in requesting such a small amount of data for a single workflow upon request would be pretty negligible compared to the delay that would happen on page load if requesting the same data for all workflows.

**Comment by alexec** [comment]
> You're correct, we're talking about the same page. I was using the term "timeline"

I guess it is call the timeline then!

I suggest lazy-loading the workflow using  `WorkflowService.get` - that way you'll have all the data you need without any problems.

I do like the idea of casually inspecting workflow like this. I think people will find it really useful.

**Comment by rbreeze** [comment]
great, sounds good! I'll start working on that. Do you think the lazy loading requires its own PR or should I include it in this one?

**Comment by rbreeze** [comment]
Upon inspection of the data returned directly from that API call, I'm finding that the response directly from the server (inspected with the Network dev tools) is exactly as expected without the additional data. However, somewhere between the server's response and the returned value of the `services.workflows.list` call, the extra data is being added somehow. 

I'm thinking there are two places this could happen: 

1. On line 27 of `workflows-service.ts` where the response is cast as a WorkflowList `return requests.get(...).then(res => res.body as WorkflowList);` This seems unlikely since `res.body` doesn't contain the data that's appearing, and at most this typecast would just assign null to the values that were missing. 

2. The subscription is somehow filling in the gaps. However, I'm not sure how this would occur either as before the subscription is unsubscribed before the API call and only resubscribed after the spot where the data is appearing (the `console.log` statement below). 

In `workflows-service.ts`:
```ts
if (this.subscription) {
    this.subscription.unsubscribe();
}
...
if (!this.state.initialized) {
    workflowList = services.info.getInfo().then(info => {
        ...
        this.setState({initialized: true, managedNamespace: !!info.managedNamespace});
        return services.workflows.list(newNamespace, selectedPhases, selectedLabels, pagination);
    });
} else {
    ...
    workflowList = services.workflows.list(newNamespace, selectedPhases, selectedLabels, pagination);
}

workflowList
.then(wfList => {
    console.log(wfList)
    ...
}).then(() => {
    this.subscription = services.workflows
        .watch({namespace: newNamespace, phases: selectedPhases, labels: selectedLabels})
        .map(workflowChange => {
             ...
```

**Comment by rbreeze** [comment]
This `console.log` statement in `workflows-service.ts` also has the extra data in it. Yet, inspecting the actual network transaction shows that the server response doesn't have the extra data. 

```ts
return requests.get(`api/v1/workflows/${namespace}?${params.join('&')}`).then(res => {
    console.log(res.body)
    return res.body as WorkflowList
});
```

Server response, from network inspector tab (the response we expect / want): 
Screen Shot 2020-06-03 at 1 16 26 PM

Output of `console.log` statement from above: 
Screen Shot 2020-06-03 at 1 17 12 PM

There are no additional requests to that endpoint, and while I suspect that the websocket is probably the culprit, I can't figure out how it would be injecting more data so soon after the return of that API call. 

Screen Shot 2020-06-03 at 1 19 23 PM

**Comment by simster7** [comment]
> great, sounds good! I'll start working on that. Do you think the lazy loading requires its own PR or should I include it in this one?

I would do the lazy loading in this PR.

You seem to have gotten a great start in debugging the Workflow object issue. Would you like to own it? If so, I'd recommend trying to fix it in a separate PR. Also, if you need help feel free to ping me

**Comment by rbreeze** [comment]
I can definitely debug the Workflow object issue, I'll open an issue and PR this afternoon. Thanks!

File: ui/src/app/shared/services/workflows-service.ts [hunk 1]
```diff
@@ -32,7 +32,11 @@ export class WorkflowsService {
     }
 
     public watch(filter: {namespace?: string; name?: string; phases?: Array<string>; labels?: Array<string>}): Observable<models.kubernetes.WatchEvent<Workflow>> {
-        const url = `api/v1/workflow-events/${filter.namespace || ''}?${this.queryParams(filter).join('&')}`;
+        const params = this.queryParams(filter);
+        params.push(
+            `fields=result.object.metadata.name,result.object.metadata.namespace,result.object.status.finishedAt,result.object.status.startedAt,result.object.status.phase`
+        );
+        const url = `api/v1/workflow-events/${filter.namespace || ''}?${params.join('&')}`;
```
Comment (simster7) [inline ui/src/app/shared/services/workflows-service.ts:39]: This seems to be fixing the `workflow-events` endpoint issue. Shouldn't this be done in #3165 instead?

File: ui/src/app/workflows/components/workflow-drawer/workflow-drawer.tsx [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+                )}
+                {!wf.status.conditions ? null : (
+                    <div className='workflow-drawer__section'>
+                        <div className='workflow-drawer__title'>CONDITIONS</div>
+                        <div className='workflow-drawer__conditions'>
+                            {wf.status.conditions.map(condition => {
+                                return (
+                                    <div className='tag' key={`${wf.metadata.namespace}-${wf.metadata.name}-${condition.type}-${condition.status}`}>
+                                        <div className='key'>{condition.type}</div>
+                                        <div className='value'>{condition.status}</div>
```
Comment (simster7) [inline ui/src/app/workflows/components/workflow-drawer/workflow-drawer.tsx:65]: With `Conditions`, the `Message` field is often more important than the status field. In fact, we only show the status field at all only if the message field is empty. Furthermore, these message fields can be quite long... take a look at this example 
image

So I'm not sure the `tag` class that we use to display "labels" would work here. What do you think?

File: ui/src/app/workflows/components/workflows-row/workflows-row.tsx [hunk 3]
```diff
@@ -6,20 +6,25 @@ import {uiUrl} from '../../../shared/base';
...
+import {services} from '../../../shared/services';
+import {WorkflowDrawer} from '../workflow-drawer/workflow-drawer';
 
 interface WorkflowsRowProps {
     workflow: models.Workflow;
     onChange: (key: string) => void;
 }
-export class WorkflowsRow extends React.Component<WorkflowsRowProps, {hideLabels: boolean}> {
+
+export class WorkflowsRow extends React.Component<WorkflowsRowProps, {hideDrawer: boolean; workflow: models.Workflow}> {
```
Comment (simster7) [inline ui/src/app/workflows/components/workflows-row/workflows-row.tsx:17]: Can we make `{hideDrawer: boolean; workflow: models.Workflow}` a format type like we do with props?

```ts
interface WorkflowRowState {
...
}
```

File: pkg/apiclient/workflow/workflow.proto [hunk 4]
```diff
@@ -89,6 +89,7 @@ message WorkflowDeleteResponse {
 message WatchWorkflowsRequest {
     string namespace = 1;
     k8s.io.apimachinery.pkg.apis.meta.v1.ListOptions listOptions = 2;
+    string fields = 3;
```
[the remark this answers is not in the document]
Comment (rbreeze) [reply pkg/apiclient/workflow/workflow.proto:92]: This was due to a gap in my understanding of git - my commits from the #3165 branch were added to this branch. I was trying to debug this enhancement with the changes I made in #3165 without considering that accordingly those changes would be added to this branch. 

To resolve this, should I simply delete the lines in question and recommit?

File: ui/src/app/workflows/components/workflow-drawer/workflow-drawer.tsx (hunk 2 above)
Comment (rbreeze) [reply ui/src/app/workflows/components/workflow-drawer/workflow-drawer.tsx:65]: Ah I see. With context, I agree - when I first saw that example, I thought that the SpecWarning portion was not itself a condition, but a constant warning displayed for every workflow.

File: pkg/apiclient/workflow/workflow.proto (hunk 4 above)
[the remark this answers is not in the document]
Comment (simster7) [reply pkg/apiclient/workflow/workflow.proto:92]: No worries. Yes, I would simply delete these lines and make sure that they are reverted exactly to what they were before by making sure they didn't change with `git diff master`

File: pkg/apiclient/workflow/workflow.pb.go [hunk 5]
```diff
@@ -806,6 +806,7 @@ var xxx_messageInfo_WorkflowDeleteResponse proto.InternalMessageInfo
 type WatchWorkflowsRequest struct {
 	Namespace            string          `protobuf:"bytes,1,opt,name=namespace,proto3" json:"namespace,omitempty"`
 	ListOptions          *v1.ListOptions `protobuf:"bytes,2,opt,name=listOptions,proto3" json:"listOptions,omitempty"`
+	Fields               string          `protobuf:"bytes,3,opt,name=fields,proto3" json:"fields,omitempty"`
```
Comment (simster7) [inline pkg/apiclient/workflow/workflow.pb.go:809]: Still need to get rid of changes in this file. Doing a `git checkout master pkg/apiclient/workflow/workflow.pb.go` should suffice

**Pull request merged**